File: typescript/server/src/choices.ts
/**
 * Choices — a structured multiple-choice ask (modeled on Claude Code's
 * `AskUserQuestion`): a **Rich Interaction kind** (see {@link ./interaction.js}).
 *
 * The agent asks 1–4 short questions, each with 2–4 labeled options; the turn
 * parks until the visitor picks. Every question also carries an implicit
 * free-text **"Other"** escape hatch, so the visitor can answer outside the
 * enumerated options (exactly as `AskUserQuestion` always offers "Other").
 *
 * - On a channel that declared the `choice_chips` capability, the agent's
 *   `request_choices` tool parks the turn and the server emits
 *   `interaction_required { kind: "choices" }`; the client's chip/menu card
 *   resumes with a `submit_interaction` action.
 * - On a **text-only** channel the same raise degrades to a conversational
 *   directive that enumerates the questions + options and the model submits the
 *   picks through the generic `submit_interaction` *tool*.
 *
 * Both paths validate through {@link validateChoices} — one implementation, one
 * behavior — and resume the turn with the same structured payload. The
 * TypeScript port of the Rust reference `smooth-operator/src/choices.rs`.
 */
import type { InteractionKind, InteractionRequest, InteractionValidation } from './interaction.js';

/** Max length of a question's short `header` label (chip/tab caption). */
export const HEADER_MAX_CHARS = 12;

/** One selectable option in a question. */
export interface ChoiceOption {
    /** The option's label — the value the visitor submits. */
    label: string;
    /** A short human-readable gloss shown under/next to the label. */
    description: string;
}

/** One question in a `choices` raise. */
export interface ChoiceQuestion {
    /** The question prompt shown to the visitor. */
    question: string;
    /** A short label (≤{@link HEADER_MAX_CHARS} chars) — the answer key and the chip/tab caption. Unique within a raise. */
    header: string;
    /** The 2–4 enumerated options. An implicit free-text "Other" is always available in addition. */
    options: ChoiceOption[];
    /** Whether the visitor may pick more than one option (default `false`). */
    multiSelect: boolean;
}

/** The visitor's answer to one question, submitted via `submit_interaction`. */
export interface ChoiceAnswer {
    /** Which question this answers — matches the spec question's `header`. */
    header: string;
    /** The selected option label(s). Empty when the visitor only used the free-text "Other". */
    options: string[];
    /** The free-text "Other" answer, when the visitor answered outside the enumerated options. Blank ⇒ omitted. */
    other?: string;
}

/** Validated, normalized choice answers — the structured payload the parked turn resumes with. */
export interface ChoiceValues {
    /** One entry per answered question (in submission order). */
    answers: ChoiceAnswer[];
}

/** A single per-question validation failure. `field` is the question's `header`. */
export interface ChoiceFieldError {
    field: string;
    message: string;
}

/** Total picks the visitor made (selected labels + one for a non-blank "Other"). */
function selectionCount(answer: ChoiceAnswer): number {
    return answer.options.length + (answer.other !== undefined ? 1 : 0);
}

/**
 * Validate submitted `values` against the raised `questions`, returning the
 * normalized {@link ChoiceValues} or the full list of per-question errors.
 *
 * Rules (mirrors `choices.rs`):
 * - **every** question must be answered (a selection or a non-blank "Other");
 * - each selected label must be one of that question's option labels;
 * - single-select: exactly one pick (one label XOR "Other"); multi-select: one or more picks;
 * - a blank/whitespace "Other" is treated as absent; labels are trimmed.
 *
 * When `questions` is empty (a prior-turn fallback raise whose spec is gone),
 * validation degrades to **format-only**: labels can't be checked for
 * membership, so any answer with at least one pick is accepted as-is.
 *
 * Returns every failed question (not just the first) so a card can annotate all
 * of them in one round-trip.
 */
export function validateChoices(questions: ChoiceQuestion[], values: ChoiceValues): { ok: true; values: ChoiceValues } | { ok: false; errors: ChoiceFieldError[] } {
    // Normalize the raw answers first (trim labels + "Other", drop blanks).
    const normalized: ChoiceAnswer[] = values.answers.map((a) => {
        const other = a.other?.trim();
        const answer: ChoiceAnswer = {
            header: a.header.trim(),
            options: a.options.map((o) => o.trim()).filter((o) => o.length > 0),
        };
        if (other && other.length > 0) answer.other = other;
        return answer;
    });

    // Format-only path: no spec to check membership / required-ness against.
    if (questions.length === 0) {
        const errors: ChoiceFieldError[] = [];
        for (const answer of normalized) {
            if (selectionCount(answer) === 0) {
                errors.push({ field: answer.header, message: "select an option or provide an 'other' answer" });
            }
        }
        if (normalized.length === 0) {
            errors.push({ field: 'answers', message: 'provide an answer for each question, or declined=true' });
        }
        return errors.length === 0 ? { ok: true, values: { answers: normalized } } : { ok: false, errors };
    }

    const errors: ChoiceFieldError[] = [];
    const out: ChoiceAnswer[] = [];

    for (const question of questions) {
        const answer = normalized.find((a) => a.header === question.header);
        if (!answer) {
            errors.push({ field: question.header, message: 'this question must be answered' });
            continue;
        }

        // Every selected label must be one of the enumerated options.
        let badLabel = false;
        for (const label of answer.options) {
            if (!question.options.some((o) => o.label === label)) {
                badLabel = true;
                errors.push({ field: question.header, message: `'${label}' is not one of the offered options` });
            }
        }

        const count = selectionCount(answer);
        if (count === 0) {
            errors.push({ field: question.header, message: "select an option or provide an 'other' answer" });
        } else if (!question.multiSelect && count > 1) {
            errors.push({ field: question.header, message: 'this question takes a single answer' });
        }

        if (!badLabel) out.push(answer);
    }

    return errors.length === 0 ? { ok: true, values: { answers: out } } : { ok: false, errors };
}

/**
 * Parse the raise tool's `questions` argument into validated {@link ChoiceQuestion}s.
 *
 * Enforces the LLM-facing contract so the model produces usable cards: 1–4
 * questions, each with a non-empty prompt, a non-empty header ≤12 chars (unique
 * within the raise), and 2–4 options with non-empty labels.
 */
export function parseQuestions(raw: unknown): ChoiceQuestion[] {
    if (!Array.isArray(raw)) throw new Error("'questions' must be an array");
    if (raw.length < 1 || raw.length > 4) throw new Error("'questions' must contain between 1 and 4 questions");

    const questions: ChoiceQuestion[] = [];
    const seenHeaders = new Set<string>();
    for (const item of raw) {
        if (typeof item !== 'object' || item === null || Array.isArray(item)) throw new Error('each question must be an object');
        const obj = item as Record<string, unknown>;

        const question = typeof obj.question === 'string' ? obj.question.trim() : '';
        if (!question) throw new Error("each question needs a non-empty 'question'");

        const header = typeof obj.header === 'string' ? obj.header.trim() : '';
        if (!header) throw new Error("each question needs a non-empty 'header'");
        if ([...header].length > HEADER_MAX_CHARS) throw new Error(`header '${header}' is too long (max ${HEADER_MAX_CHARS} characters)`);
        if (seenHeaders.has(header)) throw new Error(`duplicate question header '${header}'`);
        seenHeaders.add(header);

        const rawOptions = obj.options;
        if (!Array.isArray(rawOptions)) throw new Error(`question '${header}' needs an 'options' array`);
        if (rawOptions.length < 2 || rawOptions.length > 4) throw new Error(`question '${header}' must offer between 2 and 4 options`);

        const options: ChoiceOption[] = [];
        for (const opt of rawOptions) {
            // Accept the object form `{ label, description? }` and the shorthand bare string.
            let option: ChoiceOption;
            if (typeof opt === 'string') {
                option = { label: opt.trim(), description: '' };
            } else if (typeof opt === 'object' && opt !== null && !Array.isArray(opt)) {
                const o = opt as Record<string, unknown>;
                option = {
                    label: typeof o.label === 'string' ? o.label.trim() : '',
                    description: typeof o.description === 'string' ? o.description.trim() : '',
                };
            } else {
                throw new Error(`invalid option entry in '${header}'`);
            }
            if (!option.label) throw new Error(`an option in '${header}' has an empty label`);
            options.push(option);
        }

        questions.push({ question, header, options, multiSelect: obj.multiSelect === true });
    }
    return questions;
}

/** Coerce the submitted `values` into {@link ChoiceValues}, or throw a shape error. */
function parseValues(values: unknown): ChoiceValues {
    if (typeof values !== 'object' || values === null || Array.isArray(values)) throw new Error('values must be an object');
    const answersRaw = (values as Record<string, unknown>).answers;
    if (!Array.isArray(answersRaw)) throw new Error("values.answers must be an array");
    const answers: ChoiceAnswer[] = answersRaw.map((a) => {
        if (typeof a !== 'object' || a === null || Array.isArray(a)) throw new Error('each answer must be an object');
        const o = a as Record<string, unknown>;
        const answer: ChoiceAnswer = {
            header: typeof o.header === 'string' ? o.header : '',
            options: Array.isArray(o.options) ? o.options.filter((x): x is string => typeof x === 'string') : [],
        };
        if (typeof o.other === 'string') answer.other = o.other;
        return answer;
    });
    return { answers };
}

/**
 * The `choices` Rich Interaction kind — a structured multiple-choice ask modeled
 * on `AskUserQuestion` (see the module docs and `spec/interactions/choices.schema.json`).
 */
export class ChoicesKind implements InteractionKind {
    readonly kind = 'choices';
    readonly capability = 'choice_chips';

    toolSchema(): { name: string; description: string; parameters: Record<string, unknown> } {
        return {
            name: 'request_choices',
            description:
                'Ask the visitor a structured multiple-choice question (1–4 questions, each with 2–4 labeled options) '
                + 'and wait for their pick. On channels that can render chips/menus the visitor taps an option; on text '
                + 'channels you will be told to enumerate the options and accept a natural-language answer. An implicit '
                + 'free-text "Other" is always available, so use this whenever the answer is likely (but not certainly) '
                + 'one of a small set — never free-form the menu yourself.',
            parameters: {
                type: 'object',
                properties: {
                    questions: {
                        type: 'array',
                        minItems: 1,
                        maxItems: 4,
                        description: 'The questions to ask, in order (1–4).',
                        items: {
                            type: 'object',
                            properties: {
                                question: { type: 'string', description: 'The question prompt shown to the visitor.' },
                                header: { type: 'string', maxLength: HEADER_MAX_CHARS, description: 'A short label (≤12 chars), unique within the raise. Used as the answer key and the chip/tab caption.' },
                                options: {
                                    type: 'array',
                                    minItems: 2,
                                    maxItems: 4,
                                    description: "The 2–4 options to offer. A free-text 'Other' is always available in addition.",
                                    items: {
                                        type: 'object',
                                        properties: {
                                            label: { type: 'string', description: 'The option label (the value submitted).' },
                                            description: { type: 'string', description: 'A short gloss for the option.' },
                                        },
                                        required: ['label'],
                                    },
                                },
                                multiSelect: { type: 'boolean', description: 'Allow selecting more than one option (default false).' },
                            },
                            required: ['question', 'header', 'options'],
                        },
                    },
                    reason: {
                        type: 'string',
                        description: 'Why you\'re asking, phrased for the visitor (e.g. "to route you to the right team").',
                    },
                },
                required: ['questions', 'reason'],
            },
        };
    }

    parseRequest(args: Record<string, unknown>): InteractionRequest {
        const questions = parseQuestions(args.questions ?? null);
        const reason = typeof args.reason === 'string' && args.reason.trim().length > 0 ? args.reason.trim() : 'to help you better';
        return { kind: this.kind, spec: { questions }, reason };
    }

    validate(spec: unknown, values: unknown): InteractionValidation {
        const questions = specQuestions(spec);
        let parsed: ChoiceValues;
        try {
            parsed = parseValues(values);
        } catch (err) {
            return { ok: false, errors: [{ field: 'values', message: `invalid values shape: ${err instanceof Error ? err.message : String(err)}` }] };
        }
        const result = validateChoices(questions, parsed);
        return result.ok ? { ok: true, values: result.values } : { ok: false, errors: result.errors };
    }

    fallbackDirective(spec: unknown, reason: string): string {
        const questions = specQuestions(spec);
        const enumerated = questions
            .map((q) => {
                const opts = q.options.map((o) => o.label).join(', ');
                return `- [${q.header}] ${q.question} Options: ${opts}${q.multiSelect ? ' (choose one or more)' : ''}.`;
            })
            .join('\n');
        return (
            "This visitor's channel cannot display choice chips. Ask the following question(s) conversationally, "
            + `naturally weaving in the reason (${reason}), and read out each option so the visitor can pick:\n${enumerated}\n`
            + "The visitor may also answer with something not listed (that's fine — capture it as their 'other' answer). "
            + 'When you have their pick(s), call the `submit_interaction` tool with kind "choices" and `values.answers` — '
            + 'one entry per question `{ header, options: [chosen label(s)], other?: "their free-text answer" }`. It '
            + "validates each answer and will tell you if a pick isn't offered so you can re-ask. If the visitor declines "
            + 'to choose, call `submit_interaction` with declined=true and continue helping them.'
        );
    }
}

/** Read the `questions` out of a raise spec (best-effort; a malformed/absent spec ⇒ format-only []). */
function specQuestions(spec: unknown): ChoiceQuestion[] {
    if (typeof spec !== 'object' || spec === null) return [];
    const raw = (spec as Record<string, unknown>).questions;
    if (!Array.isArray(raw)) return [];
    const questions: ChoiceQuestion[] = [];
    for (const q of raw) {
        if (typeof q !== 'object' || q === null) continue;
        const o = q as Record<string, unknown>;
        const options = Array.isArray(o.options)
            ? o.options
                  .filter((op): op is Record<string, unknown> => typeof op === 'object' && op !== null)
                  .map((op) => ({ label: typeof op.label === 'string' ? op.label : '', description: typeof op.description === 'string' ? op.description : '' }))
            : [];
        questions.push({
            question: typeof o.question === 'string' ? o.question : '',
            header: typeof o.header === 'string' ? o.header : '',
            options,
            multiSelect: o.multiSelect === true,
        });
    }
    return questions;
}

File: go/server/choices.go
package server

import (
	"encoding/json"
	"fmt"
	"strings"
	"unicode/utf8"
)

// Choices — a structured multiple-choice ask modeled on Claude Code's
// AskUserQuestion: the reference Rich Interaction kind (the Go port of
// rust/smooth-operator/src/choices.rs). The agent asks 1–4 short questions, each
// with 2–4 labeled options; the turn parks until the visitor picks. Every question
// also carries an implicit free-text "Other" escape hatch, so the visitor can answer
// outside the enumerated options.
//
//   - On a channel that declared the `choice_chips` capability, the agent's
//     request_choices tool parks the turn and the server emits interaction_required;
//     the client's chip/menu card resumes with a submit_interaction action.
//   - On a text-only channel the same raise degrades to a conversational directive
//     enumerating the questions + options.
//
// Both paths validate through validateChoices — one implementation, one behavior —
// and resume with the same structured payload. Validated against the shared choices
// fixtures (spec/conformance/fixtures.json) in choices_test.go.

// headerMaxChars is the max length of a question's short header label (chip/tab caption).
const headerMaxChars = 12

// choiceChipsCapability gates the rich (parked card) path for the choices kind.
const choiceChipsCapability = "choice_chips"

// ChoiceOption is one selectable option in a question.
type ChoiceOption struct {
	// Label is the option's label — the value the visitor submits.
	Label string `json:"label"`
	// Description is a short human-readable gloss shown under/next to the label.
	Description string `json:"description,omitempty"`
}

// ChoiceQuestion is one question in a choices raise.
type ChoiceQuestion struct {
	// Question is the prompt shown to the visitor.
	Question string `json:"question"`
	// Header is a short label (≤headerMaxChars) — the answer key and the chip/tab
	// caption. Unique within a raise.
	Header string `json:"header"`
	// Options are the 2–4 enumerated options. An implicit free-text "Other" is always available too.
	Options []ChoiceOption `json:"options"`
	// MultiSelect reports whether the visitor may pick more than one option (default false).
	MultiSelect bool `json:"multiSelect,omitempty"`
}

// choicesSpec is the render spec carried on interaction_required for kind choices.
type choicesSpec struct {
	Questions []ChoiceQuestion `json:"questions"`
}

// ChoiceAnswer is the visitor's answer to one question, submitted via submit_interaction.
type ChoiceAnswer struct {
	// Header identifies which question this answers — matches the spec question's Header.
	Header string `json:"header"`
	// Options are the selected option label(s). One for single-select; empty when the
	// visitor only used the free-text "Other" escape hatch.
	Options []string `json:"options,omitempty"`
	// Other is the free-text "Other" answer, when the visitor answered outside the
	// enumerated options. Blank ⇒ omitted.
	Other string `json:"other,omitempty"`
}

// selectionCount is the total picks the visitor made (selected labels + one for a
// non-blank "Other").
func (a ChoiceAnswer) selectionCount() int {
	n := len(a.Options)
	if a.Other != "" {
		n++
	}
	return n
}

// ChoiceValues is the visitor's submitted answers.
type ChoiceValues struct {
	Answers []ChoiceAnswer `json:"answers"`
}

// validateChoices validates submitted values against the raised questions, returning
// the normalized ChoiceValues or the full list of per-question errors.
//
// Rules (mirror choices.rs):
//   - every question must be answered (a selection or a non-blank "Other");
//   - each selected label must be one of that question's option labels;
//   - single-select: exactly one pick (one label XOR "Other"); multi-select: ≥1;
//   - a blank/whitespace "Other" is treated as absent; labels are trimmed.
//
// When questions is empty (a prior-turn fallback raise whose spec is gone), validation
// degrades to format-only: any answer with at least one pick is accepted as-is.
// Returns every failed question (not just the first) so a card can annotate all of
// them in one round-trip.
func validateChoices(questions []ChoiceQuestion, values ChoiceValues) (ChoiceValues, []InteractionFieldError) {
	// Normalize the raw answers first (trim labels + "Other", drop blanks).
	normalized := make([]ChoiceAnswer, 0, len(values.Answers))
	for _, a := range values.Answers {
		opts := make([]string, 0, len(a.Options))
		for _, o := range a.Options {
			if t := strings.TrimSpace(o); t != "" {
				opts = append(opts, t)
			}
		}
		normalized = append(normalized, ChoiceAnswer{
			Header:  strings.TrimSpace(a.Header),
			Options: opts,
			Other:   strings.TrimSpace(a.Other),
		})
	}

	// Format-only path: no spec to check membership/required-ness against.
	if len(questions) == 0 {
		var errs []InteractionFieldError
		for _, a := range normalized {
			if a.selectionCount() == 0 {
				errs = append(errs, InteractionFieldError{Field: a.Header, Message: "select an option or provide an 'other' answer"})
			}
		}
		if len(normalized) == 0 {
			errs = append(errs, InteractionFieldError{Field: "answers", Message: "provide an answer for each question, or declined=true"})
		}
		if len(errs) > 0 {
			return ChoiceValues{}, errs
		}
		return ChoiceValues{Answers: normalized}, nil
	}

	var errs []InteractionFieldError
	out := make([]ChoiceAnswer, 0, len(questions))

	for _, q := range questions {
		answer, found := findAnswer(normalized, q.Header)
		if !found {
			errs = append(errs, InteractionFieldError{Field: q.Header, Message: "this question must be answered"})
			continue
		}

		// Every selected label must be one of the enumerated options.
		badLabel := false
		for _, label := range answer.Options {
			if !hasOption(q.Options, label) {
				badLabel = true
				errs = append(errs, InteractionFieldError{Field: q.Header, Message: fmt.Sprintf("'%s' is not one of the offered options", label)})
			}
		}

		count := answer.selectionCount()
		if count == 0 {
			errs = append(errs, InteractionFieldError{Field: q.Header, Message: "select an option or provide an 'other' answer"})
		} else if !q.MultiSelect && count > 1 {
			errs = append(errs, InteractionFieldError{Field: q.Header, Message: "this question takes a single answer"})
		}

		if !badLabel {
			out = append(out, answer)
		}
	}

	if len(errs) > 0 {
		return ChoiceValues{}, errs
	}
	return ChoiceValues{Answers: out}, nil
}

func findAnswer(answers []ChoiceAnswer, header string) (ChoiceAnswer, bool) {
	for _, a := range answers {
		if a.Header == header {
			return a, true
		}
	}
	return ChoiceAnswer{}, false
}

func hasOption(options []ChoiceOption, label string) bool {
	for _, o := range options {
		if o.Label == label {
			return true
		}
	}
	return false
}

// parseQuestions parses the raise tool's `questions` argument into validated
// ChoiceQuestions. Enforces the LLM-facing contract: 1–4 questions, each with a
// non-empty prompt, a non-empty header ≤12 chars (unique within the raise), and 2–4
// options with non-empty labels.
func parseQuestions(raw any) ([]ChoiceQuestion, error) {
	items, ok := raw.([]any)
	if !ok {
		return nil, fmt.Errorf("'questions' must be an array")
	}
	if len(items) < 1 || len(items) > 4 {
		return nil, fmt.Errorf("'questions' must contain between 1 and 4 questions")
	}
	questions := make([]ChoiceQuestion, 0, len(items))
	seen := map[string]bool{}
	for _, item := range items {
		obj, ok := item.(map[string]any)
		if !ok {
			return nil, fmt.Errorf("each question must be an object")
		}
		question := strings.TrimSpace(asString(obj["question"]))
		if question == "" {
			return nil, fmt.Errorf("each question needs a non-empty 'question'")
		}
		header := strings.TrimSpace(asString(obj["header"]))
		if header == "" {
			return nil, fmt.Errorf("each question needs a non-empty 'header'")
		}
		if utf8.RuneCountInString(header) > headerMaxChars {
			return nil, fmt.Errorf("header '%s' is too long (max %d characters)", header, headerMaxChars)
		}
		if seen[header] {
			return nil, fmt.Errorf("duplicate question header '%s'", header)
		}
		seen[header] = true

		rawOptions, ok := obj["options"].([]any)
		if !ok {
			return nil, fmt.Errorf("question '%s' needs an 'options' array", header)
		}
		if len(rawOptions) < 2 || len(rawOptions) > 4 {
			return nil, fmt.Errorf("question '%s' must offer between 2 and 4 options", header)
		}
		options := make([]ChoiceOption, 0, len(rawOptions))
		for _, opt := range rawOptions {
			// Accept the object form { label, description? } and the bare-string shorthand.
			var option ChoiceOption
			switch v := opt.(type) {
			case string:
				option = ChoiceOption{Label: strings.TrimSpace(v)}
			case map[string]any:
				option = ChoiceOption{
					Label:       strings.TrimSpace(asString(v["label"])),
					Description: strings.TrimSpace(asString(v["description"])),
				}
			default:
				return nil, fmt.Errorf("invalid option entry in '%s'", header)
			}
			if option.Label == "" {
				return nil, fmt.Errorf("an option in '%s' has an empty label", header)
			}
			options = append(options, option)
		}

		multi, _ := obj["multiSelect"].(bool)
		questions = append(questions, ChoiceQuestion{
			Question:    question,
			Header:      header,
			Options:     options,
			MultiSelect: multi,
		})
	}
	return questions, nil
}

// asString coerces a JSON-decoded value to a string ("" for anything non-string).
func asString(v any) string {
	s, _ := v.(string)
	return s
}

// ChoicesKind is the `choices` Rich Interaction kind — a structured multiple-choice
// ask modeled on AskUserQuestion (see spec/interactions/choices.schema.json).
type ChoicesKind struct{}

// Kind returns the wire kind id.
func (ChoicesKind) Kind() string { return "choices" }

// Capability returns the render capability that gates the rich card path.
func (ChoicesKind) Capability() string { return choiceChipsCapability }

// ToolSchema returns the request_choices raise tool's LLM-facing schema.
func (ChoicesKind) ToolSchema() InteractionToolSchema {
	return InteractionToolSchema{
		Name: "request_choices",
		Description: "Ask the visitor a structured multiple-choice question (1–4 questions, each " +
			"with 2–4 labeled options) and wait for their pick. On channels that can render " +
			"chips/menus the visitor taps an option; on text channels you will be told to " +
			"enumerate the options and accept a natural-language answer. An implicit free-text " +
			"\"Other\" is always available, so use this whenever the answer is likely (but not " +
			"certainly) one of a small set — never free-form the menu yourself.",
		Parameters: map[string]any{
			"type": "object",
			"properties": map[string]any{
				"questions": map[string]any{
					"type":        "array",
					"minItems":    1,
					"maxItems":    4,
					"description": "The questions to ask, in order (1–4).",
					"items": map[string]any{
						"type": "object",
						"properties": map[string]any{
							"question": map[string]any{"type": "string", "description": "The question prompt shown to the visitor."},
							"header":   map[string]any{"type": "string", "maxLength": headerMaxChars, "description": "A short label (≤12 chars), unique within the raise. Used as the answer key and the chip/tab caption."},
							"options": map[string]any{
								"type":        "array",
								"minItems":    2,
								"maxItems":    4,
								"description": "The 2–4 options to offer. A free-text 'Other' is always available in addition.",
								"items": map[string]any{
									"type": "object",
									"properties": map[string]any{
										"label":       map[string]any{"type": "string", "description": "The option label (the value submitted)."},
										"description": map[string]any{"type": "string", "description": "A short gloss for the option."},
									},
									"required": []any{"label"},
								},
							},
							"multiSelect": map[string]any{"type": "boolean", "description": "Allow selecting more than one option (default false)."},
						},
						"required": []any{"question", "header", "options"},
					},
				},
				"reason": map[string]any{"type": "string", "description": "Why you're asking, phrased for the visitor (e.g. \"to route you to the right team\")."},
			},
			"required": []any{"questions", "reason"},
		},
	}
}

// ParseRequest parses + canonicalizes the raise tool's arguments into the choices spec.
func (k ChoicesKind) ParseRequest(args map[string]any) (InteractionRequest, error) {
	questions, err := parseQuestions(args["questions"])
	if err != nil {
		return InteractionRequest{}, err
	}
	reason := strings.TrimSpace(asString(args["reason"]))
	if reason == "" {
		reason = "to help you better"
	}
	spec, err := json.Marshal(choicesSpec{Questions: questions})
	if err != nil {
		return InteractionRequest{}, err
	}
	return InteractionRequest{Kind: k.Kind(), Spec: spec, Reason: reason}, nil
}

// Validate validates submitted values against the raised spec, returning the canonical
// (normalized) values or the full list of per-question errors.
func (ChoicesKind) Validate(spec json.RawMessage, values any) (any, []InteractionFieldError) {
	var parsedSpec choicesSpec
	if len(spec) > 0 {
		// A malformed spec degrades to format-only (empty questions) rather than erroring.
		_ = json.Unmarshal(spec, &parsedSpec)
	}
	// Round-trip the decoded values (a map from the frame) through JSON into the typed shape.
	var parsedValues ChoiceValues
	rawValues, err := json.Marshal(values)
	if err == nil {
		err = json.Unmarshal(rawValues, &parsedValues)
	}
	if err != nil {
		return nil, []InteractionFieldError{{Field: "values", Message: fmt.Sprintf("invalid values shape: %v", err)}}
	}
	canonical, errs := validateChoices(parsedSpec.Questions, parsedValues)
	if len(errs) > 0 {
		return nil, errs
	}
	return canonical, nil
}

// FallbackDirective is the conversational-degradation directive for text-only channels.
func (ChoicesKind) FallbackDirective(spec json.RawMessage, reason string) string {
	var parsedSpec choicesSpec
	if len(spec) > 0 {
		_ = json.Unmarshal(spec, &parsedSpec)
	}
	lines := make([]string, 0, len(parsedSpec.Questions))
	for _, q := range parsedSpec.Questions {
		labels := make([]string, 0, len(q.Options))
		for _, o := range q.Options {
			labels = append(labels, o.Label)
		}
		multi := ""
		if q.MultiSelect {
			multi = " (choose one or more)"
		}
		lines = append(lines, fmt.Sprintf("- [%s] %s Options: %s%s.", q.Header, q.Question, strings.Join(labels, ", "), multi))
	}
	return fmt.Sprintf(
		"This visitor's channel cannot display choice chips. Ask the following question(s) "+
			"conversationally, naturally weaving in the reason (%s), and read out each option so "+
			"the visitor can pick:\n%s\nThe visitor may also answer with something not listed "+
			"(that's fine — capture it as their 'other' answer). Collect their pick(s) and continue "+
			"the conversation using the answers; one entry per question { header, options: [chosen "+
			"label(s)], other?: \"their free-text answer\" }. If a pick isn't one you offered, re-ask. "+
			"If the visitor declines to choose, continue helping them without it.",
		reason, strings.Join(lines, "\n"),
	)
}
